cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ArtifactHttpHandler.java

```diff
@@ -700,10 +698,15 @@ public BodyConsumer addArtifact(HttpRequest request, HttpResponder responder,
     } else {
       try {
         additionalPluginClasses = GSON.fromJson(pluginClasses, PLUGINS_TYPE);
+        additionalPluginClasses.forEach(PluginClass::validate);
       } catch (JsonParseException e) {
         responder.sendString(HttpResponseStatus.BAD_REQUEST, String.format(
           "%s header '%s' is invalid: %s", PLUGINS_HEADER, pluginClasses, e.getMessage()));
         return null;
+      } catch (IllegalArgumentException e) {
+        responder.sendString(HttpResponseStatus.BAD_REQUEST,
```

Comment:
I think we favor throwing exception that implements `HttpErrorStatusProvider` instead of using responder directly for error case. In this case, it's better to `throw new BadRequestException`. Same goes for the `JsonParseException` block above.